saleor/payment/utils.py

```diff
@@ -125,9 +147,91 @@ def clean_authorize(payment: Payment):
         raise PaymentError('Charged transactions cannot be authorized again.')
 
 
+def call_gateway(
+        func_name, transaction_kind, payment, payment_token, **extra_params):
+    """Helper that calls the passed gateway function and handles exceptions.
+
+    Additionally does validation of the returned gateway response."""
+    gateway, gateway_params = get_payment_gateway(payment.gateway)
+    gateway_response = None
+    error_msg = None
+
+    payment_information = create_payment_information(
+        payment, payment_token, **extra_params
+    )
+
+    try:
+        gateway_response = getattr(gateway, func_name)(
+            payment_information=payment_information, **gateway_params)
+        validate_gateway_response(gateway_response)
+    except AttributeError:
+        error_msg = 'Gateway doesn\'t implement {}'.format(func_name)
+        logger.exception(error_msg)
+    except GatewayError:
+        error_msg = 'Gateway response validation failed'
+        logger.exception(error_msg)
+        gateway_response = None  # set response empty as the validation failed
+    except Exception:
+        error_msg = 'Gateway encountered an error'
+        logger.exception(error_msg)
+    finally:
+        if not isinstance(gateway_response, list):
+            gateway_response = [gateway_response]
+        for response in gateway_response:
+            transaction = create_transaction(
+                payment=payment,
+                kind=transaction_kind,
+                payment_token=payment_token,
+                error_msg=error_msg,
+                gateway_response=response)
+
+    if not transaction.is_success:
+        # attempt to get errors from response, if none raise a generic one
+        raise PaymentError(transaction.error or GENERIC_TRANSACTION_ERROR)
+
+    return transaction
```

Comment:
Is it better that returning the last transaction explicitly if there are multiple transactions?